### PR TITLE
Support cross-compilation for iOS

### PR DESCRIPTION
**Pull request by ilammy**
There was some prior work for iOS support, but it did not really produce a successful build because running `bindgen` failed on macOS (#22). That's because bindgen tried using macOS headers instead of iOS ones, and those don't really match. Proper `-isysroot` flag needs to be passed to clang invocations made by bindgen, pointing it into the right direction.

- Fix building for all iOS targets that rustc supports:
  - `aarch64-apple-ios` – iOS devices running ARM64
  - `aarch64-apple-ios-sim` – iOS simulator running on Apple M1
  - `x86_64-apple-ios` – iOS simulator running on x86_64
- Teach CI to test building for aforementioned targets
  -  Only `cargo build`, no `cargo test`, since that's *way* more involved.
- Suppress some Clippy warnings so that the build gets green

I don't have M1 hardware so I have no idea whether `aarch64-apple-ios-sim` target actually works. Cargo says it builds though. Testing would be much appreciated.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -152,10 +152,10 @@ jobs:
     - name: Set LIBCLANG_PATH
       if: startsWith(matrix.os, 'windows')
       run: echo "LIBCLANG_PATH=$((gcm clang).source -replace "clang.exe")" >> $env:GITHUB_ENV
-    - if: startsWith(matrix.os, 'windows')
+    - if: "startsWith(matrix.os, 'windows') && !contains(matrix.target, 'ios')"
       # CI's Windows doesn't have require root certs
       run: cargo test --workspace --exclude tokio-boring --exclude hyper-boring
       name: Run tests (Windows)
-    - if: "!startsWith(matrix.os, 'windows')"
+    - if: "!startsWith(matrix.os, 'windows') && !contains(matrix.target, 'ios')"
```
Comment (jyn514) [inline .github/workflows/ci.yml:171]: Why this change? If it's because the tests were failing, why didn't they fail before this PR?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -159,3 +171,8 @@ jobs:
     - if: "!startsWith(matrix.os, 'windows') && !contains(matrix.target, 'ios')"
       run: cargo test
       name: Run tests (not Windows)
+    - if: "contains(matrix.target, 'ios')"
+      # It's... theoretically possible to run tests on iPhone Simulator,
+      # but for now, make sure that BoringSSL only builds.
+      run: cargo build --target ${{ matrix.target }}
```
Comment (jyn514) [inline .github/workflows/ci.yml:177]: This is `build` instead of check to make sure the linker succeeds?

File: boring/src/pkey.rs [hunk 3]
```diff
@@ -93,20 +93,23 @@ impl Id {
...
 /// A trait indicating that a key has public components.
+#[allow(clippy::missing_safety_doc)]
 pub unsafe trait HasPublic {}
 
 unsafe impl HasPublic for Public {}
 
 unsafe impl<T> HasPublic for T where T: HasPrivate {}
 
 /// A trait indicating that a key has private components.
+#[allow(clippy::missing_safety_doc)]
```
Comment (jyn514) [inline boring/src/pkey.rs:112]: Please move this into a separate PR, it's not related to iOS support.

**Review by jyn514 — COMMENTED** [review]
This seems reasonable, but I'm a little worried that I have no way to test it ... ideally there would be a way to cross-compile from Linux, but I know Apple makes that hard - I know some people at cloudflare with Macs, could you post instructions for how to build for iOS using a Mac?

**Comment by ilammy** [comment]
> ideally there would be a way to cross-compile from Linux, but I know Apple makes that hard

It's theoretically possible – I know people who did this some time ago with C code – but it's technically tedious (you gotta have Xcode SDKs, ideally the same clang, and *ideally* code signing setup). IIRC, it used to be legally dubious too, but nowadays Apple might be cool even if you produce binaries on non-Apple hardware. But however challenging this is, nobody publishing actual apps for iOS would go through this ordeal.

> I know some people at cloudflare with Macs, could you post instructions for how to build for iOS using a Mac?

Since `boring` is just a Rust library, it's a bit involved: you need to build a `staticlib` exporting C API for iOS app to actually use. The build step itself is simply `cargo build --target <suitable-target>`.

Here's the sample that I'm using:

https://github.com/ilammy/boring-example

That's a stub that calls SHA-256 via `boring`.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ilammy) [reply .github/workflows/ci.yml:171]: It's not possible to run `cargo test` when building for iOS. It would produce test binaries, but they can't be executed on the host macOS. You'd need to start an iOS Simulator, wrap the binaries into iOS apps, upload the apps to the simulator, run them there, extract results, and post those back. That's too much hassle. I believe it's okay to skip this step until Rust ecosystem catches up.

So for iOS builds the tests are skipped, only checking that `boring-sys` is able to compile BoringSSL and that rustc can compile Rust code that should be using the library.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (ilammy) [reply .github/workflows/ci.yml:177]: Hm... I think `check` would work fine here too, as there is no linker step. Rust does not really support producing cdylibs for iOS targets, only staticlibs, and there is no linkage there. `boring` crate itself does not provide `staticlib`, so this only builds an `rlib`, which is not really useful for building iOS targets.

File: boring/src/pkey.rs (hunk 3 above)
Comment (ilammy) [reply boring/src/pkey.rs]: I'm sorry for mixing changes. Moved this commit into #61.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (ilammy) [reply .github/workflows/ci.yml:177]: You know what, I have a better idea. How about building the tests, just `--no-run` them? Doing `cargo check` will omit all `#[cfg(test)]` code from compilation – which should be still checked by Clippy, I think, but nevertheless. Maybe some time later someone thinks up a way how to run the binaries found in the target directory.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ilammy) [reply .github/workflows/ci.yml:171]: Also, this reminds me. When CI does `cargo test`, it compiles for the host. That is, I believe that tests for Android targets are actually testing on the host Linux – not the Android 🤔

**Comment by ilammy** [comment]
I've pushed one more commit to work around https://github.com/rust-lang/rust-bindgen/issues/1651 on CI.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (jyn514) [reply .github/workflows/ci.yml]: If you want to check tests too you can use `cargo check --all-targets`.

**Comment by ilammy** [comment]
I've rebased the branch to resolve some conflicts with recently merged FIPS support.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (ilammy) [reply .github/workflows/ci.yml]: `cargo test --no-run` with `cargo check --all-targets` then.

**Comment by jyn514** [comment]
@cjmakes do you have a Mac? Can you test out https://github.com/ilammy/boring-example using the instructions in its readme?

**Comment by eager-signal** [comment]
@ilammy @jyn514 I ran through`boring-example` with the iOS Simulator on an M1 Mac and an iOS device, and both worked just fine 👍 .

**Comment by avently** [comment]
@ilammy, could you do the same magic for arm and aarch64 (Linux) target? Can't cross-compile for these platforms from x86_64 Linux host. Looks like it's not supported since https://github.com/cloudflare/quiche/blob/master/quiche/src/build.rs has specific parts related to arm/arm64 Linux while boring-sys doesn't have such parts.

I can test on aarch64 if you don't have such device.

**Review by jyn514 — APPROVED** [review]
Sorry for the long delay. Could have sworn I merged this ...
